### DISCUSSION TITLE
RFC: A unified background workers interface for Django

### DISCUSSION THREAD

**Discussion by RealOrangeOne** [django-q2/django-q2#193] [General]
I've recently submitted a Django Enhancement Proposal ("DEP") for [creating a background workers interface for Django](https://github.com/django/deps/pull/86), similar to the existing interfaces for databases and caches. The objective is to better facilitate developers (both working with Django and library maintainers) moving certain tasks from their web processes to separate background runners. Django currently has no first-party support for background workers, instead requiring developers to integrate with an external library (eg `django-q2` / `django-q`).

With a unified interface, developers will be able to more easily integrate with existing background worker libraries, without needing to re-learn a new API. Some backends will be maintained inside Django (eg some for testing, and one using its existing ORM), however the intention is for additional interfaces to integrate with existing libraries, allowing projects to use the queues they already have. 

More details, including rational and proposed API can be found in the [rendered document](https://github.com/RealOrangeOne/deps/blob/background-workers/draft/0014-background-workers.rst).

I'm opening this discussion seeking input from the `django-q2` community on the proposal's API, ensuring it doesn't hinder Django's ability to extend the API for future feature growth, but also to ensure the chosen API is compatible with a `django-q2` integration, even if said integration wouldn't expose all of `django-q2`'s features (right now). 

Feel free to reach out below, on [the related DEP PR](https://github.com/django/deps/pull/86), or directly to the contact details in the DEP description.

**Comment by matmair** [comment]
@GDay are you supportive of this proposal and have you considered accepting donations to get this supported in django-g2?
I would be intrested in (financially) backing an implementation of this as DEP 0014 has now been accepted by Django. I would like to migrate [InvenTree](https://github.com/inventree/InvenTree) to this new surface when it becomes available and django-q2 is my prefered implementation of task stuff.

**Reply by GDay** [reply]
Yeah, I guess I will have to do that at some point as this will be the standard within Django. 
I really appreciate the gesture to donate, but I don't want to turn open source work into a paid endeavor right now.

It doesn't seem to be very stable yet, so I will probably wait until this hits Django core.

**Reply by RealOrangeOne** [reply]
> It doesn't seem to be very stable yet, so I will probably wait until this hits Django core.

The integration API should be fairly stable, so you can start writing (or testing) a backend if you want to. This way issues can be found before they enter Django core.
